### PR TITLE
[Bugfix] Fix MiniCPM-o 4.5 batched Talker output attribution

### PR DESCRIPTION
**Pull request by Fyrgo8**
## Summary

- preserve the batch-aligned association between every Stage-1 request and its TTS metadata;
- run the existing blocking MiniCPM-o Talker sequentially for each valid request and emit one ordered output slot per input request;
- keep missing TTS inputs explicit with an empty float32 waveform and preserve Talker exceptions;
- make Talker dummy logits batch-shaped so a Stage-1 scheduler batch does not collapse to one row;
- add opt-in JSONL tracing for request attribution, per-stage batching/latency, TTS occupancy, queueing, GPU utilization, and peak memory;
- add a checked-in Stage-0/Stage-1 `max_num_seqs` matrix collection harness and keep all shipped defaults at `1/1`.

## Problem

`llm2tts()` already creates one Stage-1 prompt per upstream request, and the runner supplies per-request metadata in batch order. The MiniCPM-o 4.5 wrapper consumed only `runtime_additional_information[0]` and packaged at most one waveform, so a batched Stage-1 invocation could drop or misattribute later requests' outputs.

Raising `max_num_seqs` before fixing and validating this path would risk incorrect output attribution. This PR therefore fixes the batch contract first, adds deterministic coverage, and measures stage-specific settings without changing the default.

## Scope

This PR does not batch or replace the blocking HF `MiniCPMTTS.generate()` implementation, move Talker AR decoding into the vLLM scheduler, enable streaming TTS, or claim a perceptual audio-quality result. Stage-1 TTS remains sequential inside each scheduler batch.

## Batch, failure, and empty-slot semantics

- `max_num_seqs` is an admission/scheduling cap, not a guarantee that every model forward is full. A Stage-1 forward may receive several request-aligned metadata entries, but the wrapper calls the blocking `self.talker(...)` once per valid entry in a Python loop. `4/4` can therefore schedule several requests together without performing batched TTS compute.
- A real Talker exception is fail-fast. It is traced and re-raised; the wrapper stops at that slot, later slots are not executed, and no partial `OmniOutput` is returned for the scheduler batch. The exception test verifies calls stop at the failing request.
- Missing TTS input is different from a generation error. It produces an empty float32 tensor in the same output position, so worker-side payload partitioning keeps request order and cannot substitute another request's waveform. Downstream zero-length handling is endpoint-specific: realtime output emits no chunk, while non-streaming output carries zero samples. The empty tensor is a routing sentinel, not successful audio.
- All valid GPU E2E requests produced positive audio frames. The empty-slot behavior is covered by the CPU test only because normal `llm2tts()` output in these runs always contained valid TTS metadata.

## Opt-in trace format

Tracing is disabled unless `VLLM_OMNI_CONCURRENCY_TRACE_PATH` is set. Each process appends one JSON object per line with `event`, `pid`, `run_id`, wall-clock and monotonic timestamps, followed by event-specific fields. `VLLM_OMNI_CONCURRENCY_TRACE_RUN_ID` identifies one service instance.

```json
{"event":"stage1_input_ready","request_id":"chatcmpl-...-internal","batch_index":0,"batch_size":2,"has_tts_input":true,"tts_token_count":39}
{"event":"tts_slot_started","request_id":"chatcmpl-...-internal","output_slot":0}
{"event":"tts_slot_completed","request_id":"chatcmpl-...-internal","output_slot":0,"outcome":"ok","waveform_samples":52800,"latency_ms":472.1}
{"event":"request_completed","request_id":"chatcmpl-...-internal","e2e_total_ms":958.4,"transfers_total_bytes":0}
```

Other event types record the independently loaded stage configuration, changes in active model-batch request composition, stage completion/transfer times, GPU samples, and selected Prometheus queue metrics. Batch-composition events are emitted only when `(stage_id, request_ids)` changes; they support a maximum observed batch size, not a mean or per-forward distribution. The benchmark copies the trace slice for each measured group and validates that every response ID has exactly one positive `tts_slot_completed` match.

## Validation environment

- Hardware: 2x NVIDIA RTX PRO 6000 Blackwell Server Edition, 97,887 MiB each
- Placement: Stage 0 on GPU 0; Stage 1 on GPU 1
- Driver: 580.95.05
- Python: 3.12
- vLLM: 0.25.0
- vLLM-Omni package version during the GPU run: `0.1.dev2210+ga432b8a32.d20260716`
- GPU validation source: base `a432b8a3` with this PR's serving patch and collection harness applied in the working tree
- GPU validation snapshot: `88df95de07d68694980302c3f6a7617b5687e610`
- Model: MiniCPM-o 4.5, BF16, eager mode
- Warmup: 10 requests after each independent service restart
- Measurement: 50 requests per configuration/concurrency group

The GPU matrix was collected before the validation changes were committed, so the package version above records the base commit. The Talker serving-path files used for the GPU run were verified byte-identical to the GPU validation snapshot above. Subsequent commits reorganized history, applied formatting, and corrected trace naming and summary labels; they do not change the Talker batching behavior validated by the run. The environment emitted the existing vLLM/vLLM-Omni major/minor mismatch warning. All unit and E2E checks below nevertheless completed successfully.

## Correctness results

| Matrix | Formal requests | Successful | Exactly attributed | Warmup | Config instances |
|---|---:|---:|---:|---:|---:|
| 7 configs x concurrency 1/2/4 | 1,050 | 1,050 | 1,050 | 70/70 | 7/7 PASS |
| 3 configs x concurrency 2/4 x 3 rounds | 900 | 900 | 900 | 90/90 | 9/9 PASS |
| Total | 1,950 | 1,950 | 1,950 | 160/160 | 16/16 PASS |

In the repeated matrix, all 900 client request IDs and all 900 internal request IDs were unique. `stage1_input_ready`, `tts_slot_started`, `tts_slot_completed`, and `request_completed` normalized to the same request set in every group. All 900 waveform sample counts exactly matched the corresponding response audio frame counts. No server or sampler fatal pattern was found.

## Stage-specific matrix

One full matrix varied Stage-0 and Stage-1 independently. Values below are completed requests/s; the last column is the largest observed Stage-0/Stage-1 batch size at request concurrency 4.

| Stage-0/Stage-1 | c=1 | c=2 | c=4 | p95 at c=2 | p95 at c=4 | Observed batch max |
|---|---:|---:|---:|---:|---:|---:|
| 1/1 | 1.544 | 2.056 | 2.037 | 1,043 ms | 2,108 ms | 1/1 |
| 1/2 | 1.520 | 1.683 | 2.006 | 1,263 ms | 2,075 ms | 1/1 |
| 1/4 | 1.518 | 1.743 | 1.821 | 1,240 ms | 2,301 ms | 1/3 |
| 2/1 | 1.550 | 2.112 | 2.079 | 1,028 ms | 1,993 ms | 2/1 |
| 2/2 | 1.494 | 1.691 | 1.985 | 1,257 ms | 2,111 ms | 2/1 |
| 4/1 | 1.497 | 2.058 | 2.030 | 1,050 ms | 2,038 ms | 4/1 |
| 4/4 | 1.537 | 1.759 | 1.872 | 1,197 ms | 2,341 ms | 4/3 |

The apparent single-run `2/1` lead was small enough to require an order-rotated repeat rather than being treated as a configuration recommendation.

## Repeated Stage-0 ablation

The follow-up fixed Stage 1 at `1`, tested request concurrency 2 and 4, and restarted the service for every configuration. Configuration order was rotated across rounds:

1. `1/1 -> 2/1 -> 4/1`
2. `4/1 -> 1/1 -> 2/1`
3. `2/1 -> 4/1 -> 1/1`

| Config | Concurrency | RPS mean +/- sample SD | p50 mean | p95 mean | Stage-1 queue p95 mean |
|---|---:|---:|---:|---:|---:|
| 1/1 | 2 | 2.0074 +/- 0.0084 | 966 ms | 1,118 ms | 429 ms |
| 2/1 | 2 | 2.0427 +/- 0.0132 | 947 ms | 1,106 ms | 451 ms |
| 4/1 | 2 | 1.9885 +/- 0.0498 | 984 ms | 1,096 ms | 430 ms |
| 1/1 | 4 | 2.0526 +/- 0.0155 | 1,916 ms | 2,025 ms | 1,367 ms |
| 2/1 | 4 | 2.0602 +/- 0.0521 | 1,903 ms | 2,045 ms | 1,396 ms |
| 4/1 | 4 | 1.9806 +/- 0.0692 | 1,997 ms | 2,122 ms | 1,461 ms |

Paired RPS changes against `1/1`, using a Student-t 95% CI with three rounds:

| Comparison | Per-round change | Mean | 95% CI |
|---|---|---:|---:|
| `2/1` vs `1/1`, c=2 | +2.99%, +0.91%, +1.39% | +1.76% | [-0.94%, +4.46%] |
| `2/1` vs `1/1`, c=4 | +4.04%, -2.40%, -0.48% | +0.39% | [-7.82%, +8.60%] |
| `4/1` vs `1/1`, c=2 | -0.24%, +1.08%, -3.66% | -0.94% | [-7.01%, +5.13%] |
| `4/1` vs `1/1`, c=4 | -5.10%, -0.30%, -5.15% | -3.52% | [-10.44%, +3.41%] |

## Recommendation path

The recommendation applies two separate controls: the shipped Stage-0/Stage-1 `max_num_seqs` values and the request-side operating concurrency. It does not treat the numerically fastest single cell as sufficient evidence for a default change.

1. **Correctness and resource gate.** A candidate must complete every request with exact waveform attribution and no cross-request leakage, and it must avoid OOM under the measured workload. All settings passed the correctness requirement: 1,950/1,950 formal requests were successful and exactly attributed. No run OOMed on the 2x RTX PRO 6000 host. With one stage assigned to each GPU, the maximum sampled NVML device `memory.used` at 0.5-second intervals was 57.88 GiB on the Stage-0 GPU for `1/1` and 47.02 GiB for both `2/1` and `4/1`; the Stage-1 GPU stayed within 6.13-6.14 GiB. These are device-level samples, not PyTorch allocator peaks, and are specific to this hardware and workload rather than a general no-regression claim.
2. **Independent stage screen.** The seven-config matrix varied Stage 0 and Stage 1 separately. Stage-1 values above 1 did not expose batched TTS compute: scheduler batches were still consumed by sequential Talker calls and generally reduced throughput or increased latency. This rules out recommending `1/2`, `1/4`, `2/2`, or `4/4` for the current implementation.
3. **Repeat any small apparent lead.** The broad matrix showed a small one-run `2/1` lead, so `1/1`, `2/1`, and `4/1` were repeated three times with rotated order. `2/1` at c=2 was directionally positive in all three rounds, but its paired 95% CI crossed zero; at c=4 the sign was inconsistent. `4/1` showed no stable benefit.
4. **Require a mechanism, not only a number.** Stage-0 batches larger than 1 were observed: `2/1` reached batch 2 and `4/1` reached batch 4 at request concurrency 4. The trace event records changes in `(stage_id, request_ids)`, not every model forward, so it supports these maximum observed batch sizes but not a mean-batch or forward-frequency claim. Stage-0 generation p50 stayed within 145.5-147.3 ms, and cross-stage overlap at c=4 decreased from 29.0% (`1/1`) to 28.1% (`2/1`) and 26.3% (`4/1`). Combined with the confidence intervals crossing zero and the blocking TTS bottleneck, the data does not support attributing the small `2/1` c=2 difference to Stage-0 batching.
5. **Check the actual bottleneck.** Blocking TTS occupied about 98.5% of each measured wall-time window. Across the 18 repeated groups, RPS correlated strongly with TTS p50 (`r=-0.964`) but not with Stage-0 generation p50 (`r=0.079`). Mean GPU utilization was also non-monotonic:

   | Config | GPU 0 / GPU 1 at c=2 | GPU 0 / GPU 1 at c=4 |
   |---|---:|---:|
   | 1/1 | 21.2% / 42.3% | 22.4% / 43.7% |
   | 2/1 | 23.8% / 44.1% | 21.6% / 44.8% |
   | 4/1 | 21.0% / 41.5% | 21.1% / 44.0% |

   These small, inconsistent utilization differences do not explain a throughput scaling effect. The measured memory values are reported in the correctness and resource gate above and are not generalized beyond this environment.
6. **Choose request concurrency by throughput/latency tradeoff.** Raising request concurrency from 2 to 4 did not materially increase throughput, but approximately doubled p50 latency and increased Stage-1 queue p95 from roughly 0.43 s to 1.37-1.46 s.

**Recommendation:** keep the shipped Stage-0/Stage-1 default at `1/1`. In this measured two-GPU, fixed-short-prompt workload, request concurrency 2 provided the practical throughput/latency operating point; this request-side observation is not proposed as a general serving default. `2/1` at request concurrency 2 is a directionally positive observation worth retesting: it was faster than `1/1` in all three c=2 rounds, averaging +1.76% RPS with slightly lower mean p50/p95 latency. However, the limited sample size, paired 95% CI crossing zero, lack of reproduction at c=4, and absence of evidence that the observed Stage-0 batching caused the small throughput difference mean it is not recommended as either a new default or an opt-in performance setting in this PR. Stage-1 settings above 1 and `4/1` are also not recommended for the current blocking TTS path. Revisit `2/1` after additional independent repetitions or after true batched or non-blocking TTS generation removes the current bottleneck.

## Benchmark artifacts

- [Browse the sanitized aggregate artifacts](https://gist.github.com/Fyrgo8/1d7e58f191537bc4a929e6050eb66806)
- [Download the pinned artifact snapshot as ZIP](https://gist.github.com/Fyrgo8/1d7e58f191537bc4a929e6050eb66806/archive/ac510623157c424a643dbde17c67df7010a73800.zip)

The snapshot contains `aggregate_rows.csv`, `ablation_analysis.json`,
`integrity_report.json`, `mechanism_analysis.json`, `README.md`, and
`SHA256SUMS`. The link is pinned to Gist commit `ac510623`, and `SHA256SUMS`
covers the extracted files.

The snapshot predates the trace-label correction. Its `stage0_mean_model_batch` and batch-histogram fields summarize composition-change events, not model-forward frequency, and should not be used for mean-batch or "most forwards" conclusions. Maximum observed batch sizes remain valid.

Raw per-request responses, audio payloads, JSONL traces, and server logs are
intentionally excluded because they are large, environment-specific, and may
contain request content. The reproduction commands below regenerate the raw
per-group collection process; the pinned snapshot contains the offline aggregate
reports used for the tables above.

## Tests

```text
8 passed, 15 warnings in 1.01s
```

The Talker tests use a deterministic CPU-only `_RecordingTalker`; they require no model download or GPU. Marker-valued waveforms make request order and attribution directly observable. Covered cases include distinct and missing TTS metadata, ordered waveform attribution, fail-fast exception propagation (including proving later slots are not executed), no cross-invocation leakage, intermediate-buffer precedence, batch-shaped Talker logits, trace disablement, and trace aggregation.

```bash
pytest -s -v \
  tests/metrics/test_concurrency_trace.py \
  tests/model_executor/models/minicpmo_4_5/test_talker_batch.py \
  -m "core_model and cpu"
```

## Reproduction: collection

```bash
export MODEL=/path/to/MiniCPM-o-4_5
export PYTHON=/path/to/vllm-omni-env/bin/python
export VLLM=/path/to/vllm-omni-env/bin/vllm
export WARMUP_REQUESTS=10
export WARMUP_CONCURRENCY=4
export REQUESTS_PER_GROUP=50

CONCURRENCIES="1 2 4" OUTPUT_ROOT=/path/to/results/full-matrix \
  bash benchmarks/minicpmo/run_concurrency_matrix.sh

CONCURRENCIES="2 4" OUTPUT_ROOT=/path/to/results/round1 \
  bash benchmarks/minicpmo/run_concurrency_matrix.sh max_seqs_1_1 max_seqs_2_1 max_seqs_4_1
CONCURRENCIES="2 4" OUTPUT_ROOT=/path/to/results/round2 \
  bash benchmarks/minicpmo/run_concurrency_matrix.sh max_seqs_4_1 max_seqs_1_1 max_seqs_2_1
CONCURRENCIES="2 4" OUTPUT_ROOT=/path/to/results/round3 \
  bash benchmarks/minicpmo/run_concurrency_matrix.sh max_seqs_2_1 max_seqs_4_1 max_seqs_1_1
```

Relates to #5069.

### CODE REVIEW

**Comment by lishunyang12** [comment]
Which GPU you are trying to test it on?

**Comment by Fyrgo8** [comment]
@lishunyang12 I’m planning to test on 2× NVIDIA RTX PRO 6000 Blackwell 96GB GPUs, with the Thinker and Talker on separate GPUs.

**Comment by Fyrgo8** [comment]
Hi @amy-why-3459 @Sy0307,

A quick ping for a maintainer review on this PR when you have a chance. This change addresses MiniCPM-o 4.5 batched Talker output attribution and is part of #5069.

Thanks!
